core/src/main/java/org/apache/calcite/jdbc/SimpleCalciteSchema.java

```diff
@@ -67,31 +69,67 @@ public CalciteSchema add(String name, Schema schema) {
     return calciteSchema;
   }
 
+  private String caseInsensitiveLookup(Set<String> candidates, String name) {
+    // Exact string lookup
+    if (candidates.contains(name)) {
+      return name;
+    }
+    // Upper case string lookup
+    final String upperCaseName = name.toUpperCase(Locale.ROOT);
+    if (candidates.contains(upperCaseName)) {
+      return upperCaseName;
+    }
+    // Lower case string lookup
+    final String lowerCaseName = name.toLowerCase(Locale.ROOT);
+    if (candidates.contains(lowerCaseName)) {
+      return lowerCaseName;
+    }
+    // Fall through: Set iteration
+    for (String candidate: candidates) {
+      if (candidate.equalsIgnoreCase(name)) {
+        return candidate;
+      }
+    }
+    return null;
+  }
+
   protected CalciteSchema getImplicitSubSchema(String schemaName,
       boolean caseSensitive) {
     // Check implicit schemas.
-    Schema s = schema.getSubSchema(schemaName);
-    if (s != null) {
-      return new SimpleCalciteSchema(this, s, schemaName);
+    final String schemaName2 = caseSensitive ? schemaName : caseInsensitiveLookup(
+        schema.getSubSchemaNames(), schemaName);
+    if (schemaName2 != null) {
+      final Schema s = schema.getSubSchema(schemaName2);
```

Comment:
Can you please use fast return here?
That is `if (... == null) { return null }`

Reply:
Is the following code style more efficient?
```
    if (schemaName2 == null) {
      return null;
    }
    final Schema s = schema.getSubSchema(schemaName2);
    if (s == null) {
      return null;
    }
    return new SimpleCalciteSchema(this, s, schemaName2);
```

Reply:
The efficiency does not matter here.

I think we should refrain from deep {.. {..} } structures.

Reply:
Ok, thanks for clarification.

Reply:
However, it looks like this case equally bad in both styles :(

Reply:
It seems to be difficult to achieve both efficiency and elegant stlye. Maybe some functional languages (e.g., Scala) provide sugar :).